In [1]:
# Exercise 1.1

temps = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]

fahrenheit = [t * 9/5 + 32 for t in temps]

print(fahrenheit)

[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


In [2]:
# Exercise 1.2

hot = [t for t in temps if t >= 26]

cool_rounded = [round(t) for t in temps if t < 24]

print("Hot:", hot)
print("Cool rounded:", cool_rounded)

Hot: [28.6, 26.0]
Cool rounded: [19, 22]


In [3]:
# Exercise 1.3

raw = ["24.4", "25.9", "28.6", "19.3"]

raw_floats = [float(x) for x in raw]

print("Floats:", raw_floats)
print("Average:", sum(raw_floats) / len(raw_floats))

Floats: [24.4, 25.9, 28.6, 19.3]
Average: 24.55


In [4]:
# Exercise 2.1

results = [
    ("Aline", 85),
    ("Eric", 78),
    ("Diane", 91),
    ("Samuel", 92),
    ("Grace", 76)
]

ranked = sorted(results, key=lambda x: x[1], reverse=True)

print(ranked[:3])

[('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [5]:
# Exercise 2.2

rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]

wettest = max(rain_log, key=lambda x: x[1])

print("Wettest day:", wettest)

Wettest day: (14, 26.2)


In [1]:
# Exercise 3.1

import numpy as np
import time

arr = np.arange(1000000)

start = time.time()

result = []
for x in arr:
    result.append(x * 2)

loop_time = time.time() - start

start = time.time()

result_numpy = arr * 2

numpy_time = time.time() - start

print("Python loop time:", loop_time)
print("NumPy time:", numpy_time)
print("NumPy was approximately", loop_time / numpy_time, "times faster")

Python loop time: 0.5756733417510986
NumPy time: 0.011816024780273438
NumPy was approximately 48.71971347861179 times faster


# NumPy was approximately 48.7 times faster on my machine.

In [5]:
import pandas as pd
import numpy as np

df = pd.read_excel("week5_kigali_weather.xlsx")

dates = pd.to_datetime(df["date"], unit="D", origin="1899-12-30").to_numpy()
temps = df["temp_c"].to_numpy()
rain = df["rainfall_mm"].to_numpy()
hum = df["humidity_pct"].to_numpy()

In [3]:
# Exercise 3.2

print(f"Mean: {temps.mean():.2f} °C")
print(f"Max: {temps.max():.2f} °C")
print(f"Min: {temps.min():.2f} °C")
print(f"Std: {temps.std():.2f} °C")

Mean: 24.00 °C
Max: 28.60 °C
Min: 19.30 °C
Std: 2.61 °C


In [6]:
# Exercise 3.3

position = np.argmax(temps)

print("Hottest day:", dates[position])
print("Temperature:", temps[position], "°C")

Hottest day: 2026-09-17T00:00:00
Temperature: 28.6 °C


In [7]:
# Exercise 4.1

hot_mask = temps >= 26

print("Hot days:", hot_mask.sum())
print("Average hot temperature:", temps[hot_mask].mean())

Hot days: 18
Average hot temperature: 27.11111111111111


In [8]:
# Exercise 4.2

rain_mask = rain > 0

print("Rainy days:", rain_mask.sum())
print("Total rainfall:", rain.sum(), "mm")
print("Heaviest rainfall:", rain.max(), "mm")

Rainy days: 16
Total rainfall: 331.20000000000005 mm
Heaviest rainfall: 26.2 mm


In [9]:
# Exercise 4.3

rainy_mask = rain > 0
dry_mask = rain == 0

print("Average rainy temperature:", temps[rainy_mask].mean())
print("Average dry temperature:", temps[dry_mask].mean())

hot_rainy = (temps >= 26) & (rain > 0)
print("Hot and rainy days:", hot_rainy.sum())

humid_days = hum >= 70
print("Days with humidity >= 70%:", humid_days.sum())

Average rainy temperature: 24.137500000000003
Average dry temperature: 23.94545454545455
Hot and rainy days: 5
Days with humidity >= 70%: 16


In [10]:
# Exercise 5.1

anomaly = temps - temps.mean()

above_mask = anomaly > 3
below_mask = anomaly < -3

print("Dates more than 3°C above mean:")
print(dates[above_mask])

print("Days more than 3°C below mean:", below_mask.sum())

Dates more than 3°C above mean:
['2026-09-07T00:00:00' '2026-09-09T00:00:00' '2026-09-11T00:00:00'
 '2026-09-12T00:00:00' '2026-09-14T00:00:00' '2026-09-16T00:00:00'
 '2026-09-17T00:00:00' '2026-09-19T00:00:00' '2026-09-22T00:00:00']
Days more than 3°C below mean: 8


In [11]:
# Exercise 5.2

first_30_avg = temps[:30].mean()
last_30_avg = temps[30:].mean()

print(f"First 30 days average: {first_30_avg:.2f} °C")
print(f"Last 30 days average: {last_30_avg:.2f} °C")

# Kigali got warmer if the last 30 days have a higher average temperature.

First 30 days average: 26.21 °C
Last 30 days average: 21.78 °C


# Kigali got cooler.

In [12]:
# Top 5 wettest days

wettest = sorted(zip(dates, rain), key=lambda x: x[1], reverse=True)

for day, rainfall in wettest[:5]:
    print(day, rainfall, "mm")

2026-09-07T00:00:00 26.2 mm
2026-09-14T00:00:00 25.1 mm
2026-09-21T00:00:00 24.0 mm
2026-09-05T00:00:00 22.9 mm
2026-09-28T00:00:00 22.9 mm


## Part 6 — Reflection

### 1. What did NumPy hide from you, and why was it important to learn the loop version first?

NumPy hides the loops and calculations behind simple array operations such as `mean()`, `max()`, and `min()`. Learning the loop version first helped me understand what happens step by step before using NumPy to make the process faster and easier.

### 2. One insight about Kigali's weather

The first 30 days had an average temperature of 26.21°C, while the last 30 days averaged 21.78°C. This shows that Kigali became cooler during the second half of the period.